src/Storages/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,485 @@
+    {
+        LOG_ERROR(log, "Failed to allocate local file for remote path: {}, reason: {}.", remote_path, error);
+        // read_buffer is the input one.
+        return read_buffer;
+    }
+    else
+    {
+        remote_read_buffer->file_buffer = remote_read_buffer->file_cache_controller->allocFile();
+        if (!remote_read_buffer->file_buffer)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Create file readbuffer failed. {}",
+                    remote_read_buffer->file_cache_controller->getLocalPath().string());
+    }
+    remote_read_buffer->remote_file_size = remote_file_metadata->getFileSize();
+    return remote_read_buffer;
+}
+
+bool RemoteReadBuffer::nextImpl()
+{
+    auto start_offset = file_buffer->getPosition();
+    auto end_offset = start_offset + file_buffer->internalBuffer().size();
+    file_cache_controller->waitMoreData(start_offset, end_offset);
+
+    auto status = file_buffer->next();
+    if (status)
+    {
+        BufferBase::set(file_buffer->buffer().begin(),
+                file_buffer->buffer().size(),
+                file_buffer->offset());
+        ProfileEvents::increment(ProfileEvents::ExternalDataSourceLocalCacheReadBytes, file_buffer->available());
+    }
+    return status;
+}
+
+off_t RemoteReadBuffer::seek(off_t offset, int whence)
+{
+    if (!file_buffer)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot call seek() in this buffer. It's a bug!");
+    /*
+     * Need to wait here. For example, the current file has been download at position X, but here we try to seek to
+     * postition Y (Y > X), it would fail.
+     */
+    file_cache_controller->waitMoreData(offset, offset + file_buffer->internalBuffer().size());
+    auto ret = file_buffer->seek(offset, whence);
+    BufferBase::set(file_buffer->buffer().begin(),
+            file_buffer->buffer().size(),
+            file_buffer->offset());
+    return ret;
+}
+
+off_t RemoteReadBuffer::getPosition()
+{
+    return file_buffer->getPosition();
+}
+
+RemoteReadBufferCache::RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache::~RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache & RemoteReadBufferCache::instance()
+{
+    static RemoteReadBufferCache instance;
+    return instance;
+}
+
+void RemoteReadBufferCache::recoverCachedFilesMetadata(
+    const fs::path & current_path,
+    size_t current_depth,
+    size_t max_depth)
+{
+    if (current_depth >= max_depth)
+    {
+        std::vector<fs::path> invalid_pathes;
+        for (auto const & dir : fs::directory_iterator{current_path})
+        {
+            String path = dir.path();
+            auto cache_controller = RemoteCacheController::recover(path);
+            if (!cache_controller)
+            {
+                invalid_pathes.emplace_back(path);
+                continue;
+            }
+            if (!lru_caches->set(path, cache_controller))
+            {
+                invalid_pathes.emplace_back(path);
```

Comment:
what is `pathes`? probably you meant `paths`

Reply:
yes